File: LeetCodes/sum-root-to-leaf-numbers/1-15-2023, 7_14_33 PM/sum-root-to-leaf-numbers.cpp
// Link: https://leetcode.com/problems/sum-root-to-leaf-numbers

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: sum-root-to-leaf-numbers
 *  RUNTIME: 7 ms
 *  MEMORY: 9.3 MB
 *  DATE: 1-15-2023, 7:14:33 PM
 *
 */

// Solution:

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = 0;
    
    void dfs(TreeNode* node, int num) {
        if(node == NULL) return;
        
        num = num*10 + node->val;
        if(node->left == NULL && node->right == NULL) ans += num;
        
        dfs(node->left, num);
        dfs(node->right, num);
    }

    int sumNumbers(TreeNode* root) {
        dfs(root, 0);
        return ans;
    }
};

File: LeetCodes/sum-root-to-leaf-numbers/4-15-2024, 11_31_19 AM/sum-root-to-leaf-numbers.java
// Link: https://leetcode.com/problems/sum-root-to-leaf-numbers

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: sum-root-to-leaf-numbers
 *  RUNTIME: 3 ms
 *  MEMORY: 40.9 MB
 *  DATE: 4-15-2024, 11:31:19 AM
 *
 */

// Solution:

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public static int sum;
    public int sumNumbers(TreeNode root) {
        int num = 0;
        sum = 0;
        dfs(num, root);
        return sum;
    }

    public void dfs(int num, TreeNode node) {
        if(node == null) return;
        num = num * 10 + node.val;

        if(node.right == null && node.left == null) {
            System.out.println(num);
            sum += num;
            return;
        }

        dfs(num, node.right);
        dfs(num, node.left);
    }
}
